src/irc/core/irc-servers.c

```diff
@@ -603,7 +603,7 @@ static int server_cmd_timeout(IRC_SERVER_REC *server, gint64 now)
 
 	/* send command */
 	len = strlen(cmd);
-	irc_server_send_data(server, cmd, len);
+	signal_emit("server outgoing", 3, server, cmd, len);
 
 	/* add to rawlog without [CR+]LF */
```

Comment:
just looking at this patch after @JustAnotherArchivist pinged me on IRC, I notice that the wrong command will be logged to rawlog as far as I can tell

Reply:
Oh, yes. Good catch, thank you! There could be signal handlers both before and after `irc_server_send_data` that might modify `cmd`, so this issue exists on both the direct sends (where the rawlog is currently written before the signal) and the ones from the queue (where it's after). I think I'll just move the`rawlog_output` call to `irc_server_send_data`.

Should a command be in the rawlog if `net_sendbuffer_send` fails?

Reply:
While finally looking into this, I realised that it's even trickier. There's also the `server_redirect_command` business. I don't fully understand what it does yet, but it complicates things. It isn't possible to simply use `server->cmdqueue->next->data` in `irc_server_send_data` because signal handlers might change the queue inbetween (e.g. with calls to `Server::send_raw_now` in Perl), so the beginning of the queue might no longer correspond to the message being sent. The easiest way around it would be to add the redirect argument to the signal, but as I don't understand what the redirects do in the first place, I'm not sure if that would actually be a good idea.

Reply:
we had a brief talk about it on IRC and in summary we found that the current Irssi architecture does not seem to allow us to pass some variables along a signal emission unless they are part of the signal itself (or as peksy C globals). And we still don't know how to solve this issue here